bitcoin/script/miniscript.h

```diff
@@ -1155,44 +1156,46 @@ inline NodeRef<Key> DecodeScript(I& in, I last, const Ctx& ctx) {
                 in += 2;
                 to_parse.emplace_back(DecodeContext::ThreshW, 0, k);
             }
-            // a: wrapper
-            else if (in[0].first == OP_FROMALTSTACK) {
-                ++in;
-                to_parse.emplace_back(DecodeContext::Alt, -1, -1);
-                to_parse.emplace_back(DecodeContext::MaybeAndV, -1, -1);
-                to_parse.emplace_back(DecodeContext::MaybeSwap, -1, -1);
-                to_parse.emplace_back(DecodeContext::Expression, -1, -1);
-            }
             // OP_ENDIF can be WRAP_J, WRAP_D, ANDOR, OR_C, OR_D, or OR_I
             else if (in[0].first == OP_ENDIF) {
                 ++in;
                 to_parse.emplace_back(DecodeContext::EndIf, -1, -1);
                 to_parse.emplace_back(DecodeContext::MaybeAndV, -1, -1);
-                to_parse.emplace_back(DecodeContext::MaybeSwap, -1, -1);
                 to_parse.emplace_back(DecodeContext::Expression, -1, -1);
             }
             // and_b
             else if (in[0].first == OP_BOOLAND) {
                 ++in;
                 to_parse.emplace_back(DecodeContext::AndB, -1, -1);
                 to_parse.emplace_back(DecodeContext::Expression, -1, -1);
-                to_parse.emplace_back(DecodeContext::MaybeSwap, -1, -1);
-                to_parse.emplace_back(DecodeContext::Expression, -1, -1);
+                to_parse.emplace_back(DecodeContext::WExpression, -1, -1);
             }
             // or_b
             else if (in[0].first == OP_BOOLOR) {
                 ++in;
                 to_parse.emplace_back(DecodeContext::OrB, -1, -1);
                 to_parse.emplace_back(DecodeContext::Expression, -1, -1);
-                to_parse.emplace_back(DecodeContext::MaybeSwap, -1, -1);
-                to_parse.emplace_back(DecodeContext::Expression, -1, -1);
+                to_parse.emplace_back(DecodeContext::WExpression, -1, -1);
             }
             else {
                 // Unrecognised expression
                 return {};
             }
             break;
         }
+        case DecodeContext::WExpression: {
+            // a: wrapper
+            if (in >= last) return {};
+            if (in[0].first == OP_FROMALTSTACK) {
+                ++in;
+                to_parse.emplace_back(DecodeContext::Alt, -1, -1);
+            } else {
+                to_parse.emplace_back(DecodeContext::Swap, -1, -1);
+            }
+            to_parse.emplace_back(DecodeContext::MaybeAndV, -1, -1);
```

Comment:
Hmm, can a W contain an and_v? @sanket1729 ?

Reply:
From what I can tell, yes. For example s:X is type W if X is type Bo, and and_v is type Bo if the second argument is. So something like `s:and_v(v:older(500), sha256(H))` would be type W?

Reply:
Hmm, yes, I see no reason why it'd be illegal. But it is kind of pointless - a W is supposed to fail by putting 0 on the stack (so that e.g. a thresh or or_b can choose another branch instead), but a failure in the left side of an and_v will abort the entire script instantly.

Reply:
Agreed, I can't think of any sane use-case, but didn't think the code should disallow it if the spec does

Reply:
Yes, `and_v` cannot be a `W`. `and_v` can only be `B`, `K`, or `V`.

Reply:
@sanket1729 There is an s: or a: around it still.

Reply:
Anyway, this code is correct. "and_b(pk(A),a:and_v(v:pk(B),pk(C)))" is valid Miniscript.

Reply:
Yes, this is correct. I was answering something else

Reply:
General remark: `DecodeContext::Expression` cannot contain `and_v`, but `WExpression` can contain `and_v`. I am not a fan of current naming, but I also can't think of something better :)